File: epi_judge_java/epi/IntAsArrayMultiply.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class IntAsArrayMultiply {

    @EpiTest(testDataFile = "int_as_array_multiply.tsv")
    public static List<Integer> multiply(List<Integer> num1, List<Integer> num2) {
        if (num1.size() == 1 && num1.get(0) == 0) return List.of(0);
        if (num2.size() == 1 && num2.get(0) == 0) return List.of(0);

        List<Integer> result = new ArrayList<>();
        boolean isNegative = num1.get(0) * num2.get(0) < 0;
        num1.set(0, Math.abs(num1.get(0)));
        num2.set(0, Math.abs(num2.get(0)));

        for (int i = num2.size() - 1; i >= 0; i--) {
            int b = num2.get(i);
            int carry = 0;

            int index = num2.size() - i - 1;

            for (int j = num1.size() - 1; j >= 0; j--) {
                int a = num1.get(j);

                int mult = a * b;
                if (index == result.size()) {
                    result.add(0);
                }

                Integer prev = result.get(index);
                int sum = prev + mult + carry;
                carry = sum / 10;
                sum %= 10;
                result.set(index, sum);

                index++;
            }

            if (carry != 0) {
                result.add(carry);
            }
        }

        Collections.reverse(result);
        if (isNegative) result.set(0, result.get(0) * -1);

        return result;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IntAsArrayMultiply.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/int_as_array_multiply.cc
#include <vector>

#include "test_framework/generic_test.h"

using std::vector;

vector<int> MultiplyByNum(vector<int> num1, int num2);

vector<int> AddTo(vector<int> num1, vector<int> num2, int shift);

vector<int> Multiply(vector<int> num1, vector<int> num2) {
    int sign = num1[0] * num2[0] >= 0 ? 1 : -1;
    num1[0] = abs(num1[0]);
    num2[0] = abs(num2[0]);

    vector<int> result;
    result.push_back(0);

    for (int i = num2.size() - 1; i >= 0; --i) {
        vector<int> next = MultiplyByNum(num1, num2[i]);
        result = AddTo(result, next, num2.size() - i - 1);
    }

    result[0] *= sign;
    return result;
}

vector<int> AddTo(vector<int> num1, vector<int> num2, int shift) {
    vector<int> result;
    if (num2.size() == 1 && num2[0] == 0) {
        return num1;
    }

    for (int i = 0; i < shift; ++i) {
        num2.push_back(0);
    }

    int carry = 0;
    int index = 0;
    while (index < num1.size() || index < num2.size()) {
        int next = carry;
        if (index < num1.size()) {
            next += num1[num1.size() - 1 - index];
        }
        if (index < num2.size()) {
            next += num2[num2.size() - 1 - index];
        }

        carry = next / 10;
        result.push_back(next % 10);
        index++;
    }


    if (carry > 0) {
        result.push_back(carry);
    }

    std::reverse(result.begin(), result.end());
    return result;
}

vector<int> MultiplyByNum(vector<int> num1, int num2) {
    vector<int> result;
    if (num2 == 0) {
        result.push_back(0);
        return result;
    }

    int carry = 0;

    for (int i = num1.size() - 1; i >= 0; --i) {
        int next = num1[i] * num2 + carry;
        carry = next / 10;

        result.push_back(next % 10);
    }

    if (carry > 0) {
        result.push_back(carry);
    }

    std::reverse(result.begin(), result.end());
    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"num1", "num2"};
    return GenericTestMain(args, "int_as_array_multiply.cc",
                           "int_as_array_multiply.tsv", &Multiply,
                           DefaultComparator{}, param_names);
}
